### ISSUE TITLE
ReferenceError: weakly-referenced object no longer exists

### ISSUE THREAD

**Issue by 0x2b3bfa0** [googleapis/proto-plus-python#173]
## Environment

* macOS 10.15.7
* Python 3.9.1
* proto-plus 1.13.0

## Code

```python
import proto

class Mysid(proto.Message):
    name = proto.Field(proto.STRING, number=1)

class Hadopelagic(proto.Message):
    mysids = proto.RepeatedField(Mysid, number=1)

mysid = Mysid(name="Shrimpy")
Hadopelagic().mysids.extend([mysid])
```

Please note that this is the bare minimum required to trigger the exception: we aren't keeping any reference to the message object itself, which is being garbage-collected after calling the `extend` method, but there are [some cases](https://github.com/googleapis/proto-plus-python/issues/173#issuecomment-756493021) where this exception can be triggered with useful code.

## Error

```python
Traceback (most recent call last):
  File "/···/hadopelagic.py", line 24, in <module>
    error()
  File "/···/hadopelagic.py", line 20, in error
    variable = Hadopelagic().mysids.extend(individuals)
  File "/···/lib/python3.9/_collections_abc.py", line 1009, in extend
    self.append(v)
  File "/···/lib/python3.9/_collections_abc.py", line 988, in append
    self.insert(len(self), value)
  File "/···/lib/python3.9/site-packages/proto/marshal/collections/repeated.py", line 127, in insert
    pb_value = self._marshal.to_proto(self._pb_type, value, strict=True)
  File "/···/lib/python3.9/site-packages/proto/utils.py", line 40, in inner
    self._cached_values[fx.__name__] = fx(self)
  File "/···/lib/python3.9/site-packages/proto/marshal/collections/repeated.py", line 104, in _pb_type
    canary = copy.deepcopy(self.pb).add()
  File "/···/lib/python3.9/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/···/lib/python3.9/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/···/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/···/lib/python3.9/copy.py", line 210, in _deepcopy_tuple
    y = [deepcopy(a, memo) for a in x]
  File "/···/lib/python3.9/copy.py", line 210, in <listcomp>
    y = [deepcopy(a, memo) for a in x]
  File "/···/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/···/lib/python3.9/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/···/lib/python3.9/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/···/lib/python3.9/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/···/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/···/lib/python3.9/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/···/lib/python3.9/copy.py", line 151, in deepcopy
    copier = getattr(x, "__deepcopy__", None)
ReferenceError: weakly-referenced object no longer exists
```

## Issue

It looks like [_«there is no public-interface mechanism to determine the type of what should go in the list»_](https://github.com/googleapis/proto-plus-python/blob/ae946aa2a3b394fa31590224fcf50593bde0ccaa/proto/marshal/collections/repeated.py#L92-L93), and the most elegant way of determining the element type is by creating a deep copy of the underlying `self.pb` and adding an element to get its type:

https://github.com/googleapis/proto-plus-python/blob/ae946aa2a3b394fa31590224fcf50593bde0ccaa/proto/marshal/collections/repeated.py#L104-L105

## Suggestion

It would be nice to ask the base [protobuf](https://github.com/protocolbuffers/protobuf) project to implement the public [`GetEntryClass`](https://github.com/protocolbuffers/protobuf/blob/9647a7c2356a9529754c07235a2877ee676c2fd0/python/google/protobuf/internal/containers.py#L547-L548) method for [`RepeatedCompositeFieldContainer`](https://github.com/protocolbuffers/protobuf/blob/9647a7c2356a9529754c07235a2877ee676c2fd0/python/google/protobuf/internal/containers.py#L351-L458), as they've done with its quasi-sibling, [`ScalarMap`](https://github.com/protocolbuffers/protobuf/blob/9647a7c2356a9529754c07235a2877ee676c2fd0/python/google/protobuf/internal/containers.py#L461-L548), so the conflicting code can be replaced with this:
   ```python
   return self.pb.GetEntryClass()
   ```

**Comment by 0x2b3bfa0** [comment]
The original example is the bare minimum required to reproduce the issue, but it's also nonsensical, as the error clase discards the message object immediately without storing a reference. This new example provides a real use case where the bug can be reproduced while retaining a reference to the message object:

```python
import proto

class Creature(proto.Message):
    age = proto.Field(proto.INT32, number=1)

class Hadopelagic(proto.Message):
    creatures = proto.RepeatedField(Creature, number=1)

    @classmethod
    def from_reality(self):
        result = self(creatures=[])  # Only happens when empty.
        return result

    @classmethod
    def from_imagination(self):
        result = self(creatures=[])  # No matter how many.
        result.creatures.extend(self.from_reality().creatures)
        return result

hadopelagic = Hadopelagic().from_imagination()
```

**Comment by software-dov** [comment]
Issue has been fixed and is no longer reproducible.

**Comment by 0x2b3bfa0** [comment]
Apparently solved between `protobuf==3.14.0` and `protobuf==3.15.0rc1`
